enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go

```diff
@@ -429,6 +450,139 @@ func recordWriteMetrics(bytesWritten int64, err error, proxyRequestType string)
+// being written using zstd compression.
+type compressingWriteStream struct {
+	bspb.ByteStream_WriteServer
+
+	// if request is compressed or experimental flag `unconditionally_compress_writes` is not set.
+	useOriginalStream bool
+	bufferPool        *bytebufferpool.VariableSizePool
+	compressor        *compression.ZstdCompressingWriter
+	compressBuf       []byte
+	resourceName      string
+	writeOffset       int64
+	finishWriteSeen   bool
+}
+
+// Recv reads from the underlying stream and compresses the data.
+// It may call Recv multiple times on the underlying stream to accumulate enough
+// data to produce compressed output.
+func (c *compressingWriteStream) Recv() (*bspb.WriteRequest, error) {
+	if c.useOriginalStream { // skip on first request
+		return c.ByteStream_WriteServer.Recv()
+	}
+
+	for {
+		req, err := c.ByteStream_WriteServer.Recv()
+		if err != nil {
+			return nil, err
+		}
+
+		if c.compressor == nil {
+			if err := c.initializeCompressor(req); err != nil {
+				return nil, err
+			}
+
+			// If we decided to use the original stream, return the first request and
+			// all subsequent calls will use the original stream.
+			if c.useOriginalStream {
+				return req, nil
+			}
+		}
+
+		if len(req.GetData()) > 0 {
+			if _, err := c.compressor.Write(req.GetData()); err != nil {
+				return nil, err
+			}
+		}
+
+		if req.GetFinishWrite() {
+			c.finishWriteSeen = true
+			if err := c.compressor.Close(); err != nil {
+				return nil, err
+			}
+		}
+
+		if len(c.compressBuf) > 0 {
+			compressed := c.compressBuf
+			c.compressBuf = make([]byte, 0, compressBufSize)
+
+			outReq := &bspb.WriteRequest{
+				ResourceName: c.resourceName,
+				Data:         compressed,
+				WriteOffset:  c.writeOffset,
+				FinishWrite:  c.finishWriteSeen,
+			}
+			c.writeOffset += int64(len(compressed))
+			return outReq, nil
+		}
+
+		if c.finishWriteSeen {
+			return &bspb.WriteRequest{
+				ResourceName: c.resourceName,
+				Data:         nil,
+				WriteOffset:  c.writeOffset,
+				FinishWrite:  true,
+			}, nil
+		}
+	}
+}
+
+func (c *compressingWriteStream) initializeCompressor(req *bspb.WriteRequest) error {
+	// Check if we should skip compression
+	if shouldUseOriginalStream(req) {
+		c.useOriginalStream = true
+		return nil
+	}
+
+	c.resourceName = strings.Replace(req.GetResourceName(), "/blobs/", "/compressed-blobs/zstd/", 1)
+	var err error
+	c.compressor, err = compression.NewZstdCompressingWriter(&bufferWriter{buf: &c.compressBuf}, c.bufferPool, compressBufSize)
+	return err
+}
+
+func shouldUseOriginalStream(req *bspb.WriteRequest) bool {
+	rn, err := digest.ParseUploadResourceName(req.GetResourceName())
+	if err != nil {
+		return true
+	}
+
+	if rn.GetCompressor() == repb.Compressor_ZSTD {
+		return true
+	}
+
+	if rn.GetDigest().GetSizeBytes() < config.MinSizeToUnconditionallyCompress() {
+		return true
+	}
+
+	return false
+}
+
+type bufferWriter struct {
+	buf *[]byte
```

Comment:
might be faster to use bytes.Buffer here? Also fine to wait until profiles show this as a problem